File: tools/t_class_classifier/python/sanitize.py
from typing import Dict, Any

FORBIDDEN_FIELDS = ["C_orient", "-(i)", "𝒪", "orientation_regime", "orientation_label", "S_closure"]

def validate_schema_integrity(data: Dict[str, Any]) -> None:
    # Check if essential keys exist
    required = ["continuation_trace", "closure_adjacency", "connectivity_record"]
    for req in required:
        if req not in data:
            raise ValueError(f"Missing required field in trace JSON: '{req}'")

def strip_forbidden_fields(data: Dict[str, Any], action: str = "reject") -> Dict[str, Any]:
    # Check for forbidden fields anywhere in the dictionary keys or nested metadata
    def find_forbidden(d: Any) -> bool:
        if isinstance(d, dict):
            for k, v in d.items():
                if k in FORBIDDEN_FIELDS:
                    return True
                if find_forbidden(v):
                    return True
        elif isinstance(d, list):
            for item in d:
                if find_forbidden(item):
                    return True
        return False

    if find_forbidden(data):
        if action == "reject":
            raise ValueError(f"Forbidden field detected in input trace data: Rejecting execution.")
        elif action == "strip":
            # Recursively strip forbidden fields
            def strip_dict(d: Any) -> Any:
                if isinstance(d, dict):
                    return {k: strip_dict(v) for k, v in d.items() if k not in FORBIDDEN_FIELDS}
                elif isinstance(d, list):
                    return [strip_dict(item) for item in d]
                return d
            return strip_dict(data)
            
    return data

File: tools/t_class_classifier/cpp/sanitize.cpp
#include "json.hpp"
#include <string>
#include <vector>
#include <stdexcept>

using json = nlohmann::json;

const std::vector<std::string> FORBIDDEN_FIELDS = {
    "C_orient", "-(i)", "𝒪", "orientation_regime", "orientation_label", "S_closure"
};

bool find_forbidden_recursive(const json& j) {
    if (j.is_object()) {
        for (auto it = j.begin(); it != j.end(); ++it) {
            for (const auto& forbidden : FORBIDDEN_FIELDS) {
                if (it.key() == forbidden) {
                    return true;
                }
            }
            if (find_forbidden_recursive(it.value())) {
                return true;
            }
        }
    } else if (j.is_array()) {
        for (const auto& item : j) {
            if (find_forbidden_recursive(item)) {
                return true;
            }
        }
    }
    return false;
}

void strip_forbidden_recursive(json& j) {
    if (j.is_object()) {
        std::vector<std::string> keys_to_remove;
        for (auto it = j.begin(); it != j.end(); ++it) {
            for (const auto& forbidden : FORBIDDEN_FIELDS) {
                if (it.key() == forbidden) {
                    keys_to_remove.push_back(it.key());
                }
            }
        }
        for (const auto& k : keys_to_remove) {
            j.erase(k);
        }
        for (auto it = j.begin(); it != j.end(); ++it) {
            strip_forbidden_recursive(it.value());
        }
    } else if (j.is_array()) {
        for (auto& item : j) {
            strip_forbidden_recursive(item);
        }
    }
}

void validate_schema_integrity(const json& j) {
    std::vector<std::string> required = {"continuation_trace", "closure_adjacency", "connectivity_record"};
    for (const auto& req : required) {
        if (!j.contains(req)) {
            throw std::runtime_error("Missing required field in trace JSON: '" + req + "'");
        }
    }
}

void sanitize_and_filter(json& j, const std::string& action) {
    if (find_forbidden_recursive(j)) {
        if (action == "reject") {
            throw std::runtime_error("Forbidden field detected in input trace data: Rejecting execution.");
        } else if (action == "strip") {
            strip_forbidden_recursive(j);
        }
    }
}
